static/js/stream_active.js

```diff
@@ -0,0 +1,45 @@
+import {page_params} from "./page_params";
+import * as settings_config from "./settings_config";
+import * as stream_data from "./stream_data";
+import * as stream_topic_history from "./stream_topic_history";
+
+let filter_out_inactives = false;
+
+export function set_filter_out_inactives() {
+    if (
+        page_params.demote_inactive_streams ===
+        settings_config.demote_inactive_streams_values.automatic.code
+    ) {
+        filter_out_inactives = stream_data.num_subscribed_subs() >= 30;
+    } else if (
+        page_params.demote_inactive_streams ===
+        settings_config.demote_inactive_streams_values.always.code
+    ) {
+        filter_out_inactives = true;
+    } else {
+        filter_out_inactives = false;
+    }
+}
+
+// for testing:
+export function is_filtering_inactives() {
+    return filter_out_inactives;
+}
+
+export function is_active(sub) {
+    if (!filter_out_inactives || sub.pin_to_top) {
+        // If users don't want to filter inactive streams
+        // to the bottom, we respect that setting and don't
+        // treat any streams as dormant.
+        //
+        // Currently this setting is automatically determined
+        // by the number of streams.  See the callers
+        // to set_filter_out_inactives.
+        return true;
+    }
+    return stream_topic_history.stream_has_topics(sub.stream_id) || sub.newly_subscribed;
+}
+
+export function initialize() {
+    set_filter_out_inactives();
+}
```

Comment:
So, `active` is kinda a poor name for this bundle of stuff, because it confuses with `deactivated` streams.

Perhaps we can call this `inactive_stream_demotion.js`?  I have a hard time coming up with what to call the functions, but `stream_active` really sounded to me like some denormalized structure for telling you whether a stream has been deactivated or not.

Reply:
I would prefer a much shorter name.  The previous api was `stream_data.is_active`, which never seemed to cause confusion.  But if you decide on a final name, I'll name it whatever.

Reply:
I also considered `stream_activity`, but that would possibly cause confusion with `activity.js`.

Reply:
I think `stream_demotion.js` might be a good compromise.  Not quite so wordy, but indicates the context a bit more clearly.  The `inactive_` prefix seems entirely redundant, since you're demoting from active to inactive implicitly.

Reply:
`stream_demotion.js` works for me.